### PR TITLE
[lang] Hide internal apis about Fields

### PR DESCRIPTION
**Pull request by mzmzm**
Related issue = #3782

### CODE REVIEW

**Review by strongoier — COMMENTED** [review]
You may need to cancel modifications to `loop_range`, otherwise it will conflict with #4303.

**Comment by mzmzm** [comment]
> You may need to cancel modifications to `loop_range`, otherwise it will conflict with #4303.

Done

File: python/taichi/lang/struct.py [hunk 1]
```diff
@@ -293,60 +293,60 @@ class StructField(Field):
     def __init__(self, field_dict, name=None):
         # will not call Field initializer
         self.field_dict = field_dict
-        self._name = name
-        self.register_fields()
+        self.__name = name
```
Comment (strongoier) [inline python/taichi/lang/struct.py:296]: For this we can directly say `self.name = name` and remove the property below.

File: tests/python/test_struct_for_intermediate.py [hunk 2]
```diff
@@ -12,7 +12,7 @@ def _test_nested():
 
     @ti.kernel
     def iterate():
-        for i, j in x.parent():
+        for i, j in x._parent():
```
Comment (strongoier) [inline tests/python/test_struct_for_intermediate.py:15]: This is a huge break of user code. Let's go back to `parent()`.

File: python/taichi/lang/struct.py (hunk 1 above)
Comment (mzmzm) [reply python/taichi/lang/struct.py]: done

File: tests/python/test_struct_for_intermediate.py (hunk 2 above)
Comment (mzmzm) [reply tests/python/test_struct_for_intermediate.py]: done

File: python/taichi/lang/field.py [hunk 3]
```diff
@@ -220,6 +218,26 @@ class ScalarField(Field):
...
+        """Gets corresponding SNode for info purposes.
+
+        Returns:
+            SNode: corresponding SNode (SNode of first field member).
+        """
+        return self._snode
+
+    def parent(self, n=1):
+        """Gets an ancestor of the representative SNode in the SNode tree.
+
```
Comment (strongoier) [inline python/taichi/lang/field.py:232]: `snode` and `parent` should be moved back to the base class.

File: python/taichi/lang/matrix.py [hunk 4]
```diff
@@ -1100,6 +1100,15 @@ def __init__(self, _vars, n, m):
         self.m = m
         self.dynamic_index_stride = None
 
+    @property
+    def snode(self):
+        """Gets corresponding SNode for info purposes.
```
Comment (strongoier) [inline python/taichi/lang/matrix.py:1105]: ditto

File: python/taichi/lang/field.py [hunk 5]
```diff
@@ -44,29 +53,27 @@ def dtype(self):
...
 
         Returns:
             str: Field name.
         """
-        return self.snode._name
+        return self._snode._name
 
     def parent(self, n=1):
         """Gets an ancestor of the representative SNode in the SNode tree.
-
```
Comment (strongoier) [inline python/taichi/lang/field.py:60]: Why remove this line?

File: python/taichi/lang/field.py [hunk 6]
```diff
@@ -44,29 +53,27 @@ def dtype(self):
...
         Args:
             n (int): the number of levels going up from the representative SNode.
-
```
Comment (strongoier) [inline python/taichi/lang/field.py:63]: ditto

File: python/taichi/lang/field.py (hunk 5 above)
Comment (mzmzm) [reply python/taichi/lang/field.py:60]: fix

File: python/taichi/lang/field.py (hunk 6 above)
Comment (mzmzm) [reply python/taichi/lang/field.py:63]: fix

**Review by strongoier — APPROVED** [review]
LGTM!

**Pull request merged**